Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs

```diff
@@ -39,43 +33,66 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
         public override async Task<InnerStepResult> ExecuteAsync(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
-            context.GetNextBlock();
-
-            if (await ShouldStopFindingBlocks(context))
-            {
-                if (!context.DownloadStack.Any())
-                {
-                    this.logger.LogTrace("(-):{0}", InnerStepResult.Stop);
-                    return InnerStepResult.Stop;
-                }
 
-                context.StopFindingBlocks();
-            }
-            else
+            while (await ShouldStopFindingBlocks(context) == false)
             {
-                context.BlockStoreLoop.BlockPuller.AskBlock(context.NextChainedBlock);
                 context.DownloadStack.Enqueue(context.NextChainedBlock);
-
-                if (context.DownloadStack.Count == context.BlockStoreLoop.BatchDownloadSize)
-                    context.StopFindingBlocks();
+                context.GetNextBlock();
             }
 
+            context.StopFindingBlocks();
+
+            if (context.DownloadStack.Any())
+                context.BlockStoreLoop.BlockPuller.AskForMultipleBlocks(context.DownloadStack.ToArray());
+
             this.logger.LogTrace("(-):{0}", InnerStepResult.Next);
+
             return InnerStepResult.Next;
         }
 
         private async Task<bool> ShouldStopFindingBlocks(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
 
-            bool res = (context.NextChainedBlock == null)
-                || (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock)
-                || (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
-                || (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
-                || (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock));
+            if (context.NextChainedBlock == null)
+            {
+                this.logger.LogTrace("(-): {0} was null, return true", nameof(context.NextChainedBlock));
```

Comment:
logging the result is fine - as mentioned, you are the one who should decide which extra information is useful to have in logs

so first, you have some base line (log input parameters and output results) and then you decide if you want to add something, so if you don't see a benefit in having that context information, it is fine, although in this specific case you would end up with the code you have replaced (which is just logging the result without that context - and yes i understand that in debugger you could see more easier which case it was with your separated ifs, but still even in that joined code you could see that quite easily (just hover your mouse over the expression)) so if you will keep the ifs separated, having the context does not add any lines, just small little context identifier

but change it as you think and let me check then